File: youtube-feed-service/src/main/java/com/richo/reader/youtube_feed_service/Feed.java
package com.richo.reader.youtube_feed_service;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.richodemus.reader.dto.FeedId;
import com.richodemus.reader.dto.FeedName;

import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.List;
import java.util.Objects;

import static com.google.common.base.Preconditions.checkNotNull;

public class Feed
{
	private final FeedId id;
	private final FeedName name;
	private final List<Item> items;
	private final LocalDateTime lastUpdated;

	@JsonCreator
	public Feed(@JsonProperty("id") FeedId id,
				@JsonProperty("name") FeedName name,
				@JsonProperty("items") List<Item> items,
				@JsonProperty("lastUpdated") long lastUpdated)
	{
		this(id, name, items, LocalDateTime.ofEpochSecond(lastUpdated, 0, ZoneOffset.UTC));
	}

	public Feed(FeedId id, FeedName name, List<Item> items, LocalDateTime lastUpdated)
	{
		this.id = checkNotNull(id, "Id can't be null");
		this.name = name;
		this.items = items;
		this.lastUpdated = lastUpdated;
	}

	public FeedId getId()
	{
		return id;
	}

	public FeedName getName()
	{
		return name;
	}

	public List<Item> getItems()
	{
		return items;
	}

	@JsonIgnore
	public LocalDateTime getLastUpdated()
	{
		return lastUpdated;
	}

	@JsonProperty("lastUpdated")
	public long getLastUpdatedAsLong()
	{
		return lastUpdated.toEpochSecond(ZoneOffset.UTC);
	}

	@Override
	public boolean equals(Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		Feed feed = (Feed) o;
		return Objects.equals(id, feed.id) &&
				Objects.equals(items, feed.items);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(id, items);
	}

	@Override
	public String toString()
	{
		return name + " (" + id + ") " + items.size() + " items";
	}
}

File: reader/server/services/youtube-feed-service/src/main/kotlin/com/richodemus/reader/youtube_feed_service/Feed.kt
package com.richodemus.reader.youtube_feed_service

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import com.richodemus.reader.dto.FeedId
import com.richodemus.reader.dto.FeedName
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.util.Objects

class Feed(val id: FeedId,
           val name: FeedName,
           val items: List<Item>,
           @get:JsonIgnore private val lastUpdated: LocalDateTime) {

    @JsonCreator
    constructor(@JsonProperty("id") id: FeedId,
                @JsonProperty("name") name: FeedName,
                @JsonProperty("items") items: List<Item>,
                @JsonProperty("lastUpdated") lastUpdated: Long) :
            this(id, name, items, LocalDateTime.ofEpochSecond(lastUpdated, 0, ZoneOffset.UTC))

    @JsonProperty("lastUpdated")
    fun getLastUpdatedAsLong() = lastUpdated.toEpochSecond(ZoneOffset.UTC)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val feed = other as Feed?
        return id == feed!!.id && items == feed.items
    }

    override fun hashCode(): Int {
        return Objects.hash(id, items)
    }

    override fun toString(): String {
        return name.toString() + " (" + id + ") " + items.size + " items"
    }
}
